pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java

```diff
@@ -156,36 +158,33 @@ ElasticSearchClient getElasticsearchClient() {
             }
 
             String id = null;
-            if (key != null) {
-                if (keySchema != null) {
-                    id = stringifyKey(keySchema, key);
-                } else {
-                    id = key.toString();
-                }
+            if (elasticSearchConfig.isKeyIgnore() == false && key != null && keySchema != null) {
+                id = stringifyKey(keySchema, key);
             }
 
             String doc = null;
             if (value != null) {
                 if (valueSchema != null) {
                     doc = stringifyValue(valueSchema, value);
                 } else {
-                    doc = value.toString();
+                    if (value instanceof byte[]) {
+                        // for BWC with the ES-Sink
+                        doc = new String( (byte[]) value);
+                    } else {
+                        doc = value.toString();
+                    }
                 }
             }
 
-            if (elasticSearchConfig.isKeyIgnore()) {
-                if (doc == null || Strings.isNullOrEmpty(elasticSearchConfig.getPrimaryFields())) {
-                    // use the messageId as the doc id in last resort.
-                    id = Hex.encodeHexString(record.getMessage().get().getMessageId().toByteArray());
-                } else {
-                    try {
-                        // extract the PK from the JSON document
-                        JsonNode jsonNode = objectMapper.readTree(doc);
-                        List<String> pkFields = Arrays.asList(elasticSearchConfig.getPrimaryFields().split(","));
-                        id = stringifyKey(jsonNode, pkFields);
-                    } catch (JsonProcessingException e) {
-                        log.error("Failed to read JSON", e);
-                    }
+            if (doc != null && !Strings.isNullOrEmpty(elasticSearchConfig.getPrimaryFields())) {
+                try {
+                    // extract the PK from the JSON document
+                    JsonNode jsonNode = objectMapper.readTree(doc);
+                    List<String> pkFields = Arrays.asList(elasticSearchConfig.getPrimaryFields().split(","));
```

Comment:
we can precompute "elasticSearchConfig.getPrimaryFields().split(",")"

this way we can save a few cycles and memory allocations on this hot path

Reply:
Ok, fixed